File: ESTRUCURAS_DE_DATOS/Pilas/pythonEjercicios/polindromo.py
"En este codigo se implementa una pila con los metodos push, pop, is_empty, top y print_stack, y se utilizarad para crear un ejemplo practico y sencillo con el uso de pilas, al estilo de los anteriores ejemplos. "

class Stack:
    def __init__(self):
        self.items = []

    def push(self, item):
        self.items.append(item)

    def pop(self):
        if not self.is_empty():
            return self.items.pop()

    def is_empty(self):
        return len(self.items) == 0

    def top(self):
        if not self.is_empty():
            return self.items[-1]

    def print_stack(self):
        print(self.items)

# Ejemplo práctico: Verificar si una palabra es un palíndromo utilizando una pila
def es_palindromo(palabra):
    pila = Stack()
    for char in palabra:
        pila.push(char)

    
    print(pila.items)

    palabra_invertida = ""
    while not pila.is_empty():
        palabra_invertida += pila.pop()

    print(palabra_invertida)

    return palabra == palabra_invertida

# Prueba del ejemplo
palabra = "reconocer"
print(f"¿'{palabra}' es palíndromo?: {es_palindromo(palabra)}")

palabra2 = "pila"
print(f"¿'{palabra2}' es palíndromo?: {es_palindromo(palabra2)}")

File: ESTRUCURAS_DE_DATOS/Pilas/typescriptEjercicios/polindromo.ts
// Definición de la estructura Stack en TypeScript
class Stack<T> {
    private items: T[] = [];

    isEmpty(): boolean {
        return this.items.length === 0;
    }

    push(item: T): void {
        this.items.push(item);
    }

    pop(): T | undefined {
        if (!this.isEmpty()) {
            return this.items.pop();
        }
        return undefined;
    }

    top(): T | undefined {
        if (!this.isEmpty()) {
            return this.items[this.items.length - 1];
        }
        return undefined;
    }

    printStack(): void {
        console.log(this.items);
    }
}

// Función para verificar si una palabra/frase es palíndromo utilizando una pila
function esPalindromo(texto: string): boolean {
    const pila = new Stack<string>();
    // Limpiamos el texto: quitamos espacios y lo pasamos a minúsculas para facilitar la comprobación
    const limpio = texto.replace(/[\W_]/g, '').toLowerCase();

    // Apilamos cada caracter
    for (let char of limpio) {
        pila.push(char);
    }

    // Comparamos desapilando con el texto original limpio
    for (let char of limpio) {
        const letraPila = pila.pop();
        if (char !== letraPila) {
            return false;
        }
    }
    return true;
}

// Ejemplo de uso
const casos = [
    "Reconocer",
    "anita lava la tina",
    "Hola mundo",
    "A man, a plan, a canal: Panama"
];

casos.forEach((palabra) => {
    console.log(`'${palabra}' es palíndromo?`, esPalindromo(palabra) ? "Sí" : "No");
});
